File: template/modules/react-native-vision-engine/ios/YoloOutputDecoder.swift
import CoreML
import Foundation

/// Детекция в нормализованных координатах (top-left origin)
struct RawDetection {
  let rect: CGRect
  let score: Float
  /// Индекс класса модели; -1 — класс известен только меткой
  let classIndex: Int32
  /// Метка класса, если модель её содержит; иначе пустая строка
  let label: String
}

/// Чистый декодер сырых тензоров YOLO. Поддерживает оба поколения формата
/// выхода, различаемых по размерности:
/// - классический (v8/11/12): `[C, N]`/`[N, C]`, тысячи кандидатов
///   `cx,cy,w,h` — фильтр по score + NMS;
/// - end-to-end (v10/26): `[N, 6]` с малым N — готовые боксы
///   `x1,y1,x2,y2,score,class`.
/// Координаты нормализуются адаптивно: значения крупнее 1.5 считаются
/// пикселями входа модели.
enum YoloOutputDecoder {
  /// Верхняя граница числа детекций у end-to-end моделей (обычно 300)
  private static let maxEndToEndDetections = 512
  private static let nmsIouThreshold: CGFloat = 0.45

  /// Тензор выхода модели → детекции, отсортированные по score
  static func decode(
    _ array: MLMultiArray,
    inputSide: CGFloat,
    minConfidence: Float
  ) -> [RawDetection] {
    var dims = array.shape.map { $0.intValue }
    var strides = array.strides.map { $0.intValue }
    while dims.count > 2 && dims[0] == 1 {
      dims.removeFirst()
      strides.removeFirst()
    }
    guard dims.count == 2, let read = makeReader(array) else {
      return []
    }
    let rows = dims[0]
    let cols = dims[1]
    let rowStride = strides[0]
    let colStride = strides[1]

    func norm(_ value: Float) -> CGFloat {
      let cg = CGFloat(value)
      return cg > 1.5 ? cg / inputSide : cg
    }

    var detections: [RawDetection] = []

    if cols == 6 && rows <= maxEndToEndDetections {
      for i in 0..<rows {
        let score = read(i * rowStride + 4 * colStride)
        if score < minConfidence {
          continue
        }
        let x1 = norm(read(i * rowStride))
        let y1 = norm(read(i * rowStride + colStride))
        let x2 = norm(read(i * rowStride + 2 * colStride))
        let y2 = norm(read(i * rowStride + 3 * colStride))
        if x2 <= x1 || y2 <= y1 {
          continue
        }
        detections.append(RawDetection(
          rect: CGRect(x: x1, y: y1, width: x2 - x1, height: y2 - y1),
          score: score,
          classIndex: Int32(read(i * rowStride + 5 * colStride)),
          label: ""
        ))
      }
    } else {
      let channelsFirst = rows < cols
      let count = channelsFirst ? cols : rows
      let channels = channelsFirst ? rows : cols
      guard channels > 4 else {
        return []
      }
      func value(_ channel: Int, _ index: Int) -> Float {
        return channelsFirst
          ? read(channel * rowStride + index * colStride)
          : read(index * rowStride + channel * colStride)
      }
      for i in 0..<count {
        var score: Float = 0
        var classIndex: Int32 = 0
        for c in 4..<channels {
          let classScore = value(c, i)
          if classScore > score {
            score = classScore
            classIndex = Int32(c - 4)
          }
        }
        if score < minConfidence {
          continue
        }
        let cx = norm(value(0, i))
        let cy = norm(value(1, i))
        let w = norm(value(2, i))
        let h = norm(value(3, i))
        let x = max(cx - w / 2, 0)
        let y = max(cy - h / 2, 0)
        detections.append(RawDetection(
          rect: CGRect(x: x, y: y, width: min(w, 1 - x), height: min(h, 1 - y)),
          score: score,
          classIndex: classIndex,
          label: ""
        ))
      }
      detections = nonMaxSuppression(detections)
    }

    return detections.sorted { $0.score > $1.score }
  }

  /// Быстрое чтение MLMultiArray по плоскому индексу для float32/float16/double
  private static func makeReader(_ array: MLMultiArray) -> ((Int) -> Float)? {
    switch array.dataType {
    case .float32:
      let pointer = array.dataPointer.bindMemory(to: Float32.self, capacity: array.count)
      return { pointer[$0] }
    case .double:
      let pointer = array.dataPointer.bindMemory(to: Double.self, capacity: array.count)
      return { Float(pointer[$0]) }
    case .float16:
      #if arch(arm64)
      let pointer = array.dataPointer.bindMemory(to: Float16.self, capacity: array.count)
      return { Float(pointer[$0]) }
      #else
      return { Float(truncating: array[$0]) }
      #endif
    default:
      return nil
    }
  }

  /// Жадный NMS: кандидат с IoU выше порога к уже принятым отбрасывается
  private static func nonMaxSuppression(_ detections: [RawDetection]) -> [RawDetection] {
    let sorted = detections.sorted { $0.score > $1.score }
    var kept: [RawDetection] = []
    for candidate in sorted {
      let overlaps = kept.contains { iou($0.rect, candidate.rect) > nmsIouThreshold }
      if !overlaps {
        kept.append(candidate)
      }
    }
    return kept
  }

  private static func iou(_ a: CGRect, _ b: CGRect) -> CGFloat {
    let intersection = a.intersection(b)
    if intersection.isNull || intersection.isEmpty {
      return 0
    }
    let intersectionArea = intersection.width * intersection.height
    let unionArea = a.width * a.height + b.width * b.height - intersectionArea
    return unionArea <= 0 ? 0 : intersectionArea / unionArea
  }
}

File: template/modules/react-native-vision-engine/android/src/main/java/com/margelo/nitro/visionengine/YoloOutputDecoder.kt
package com.margelo.nitro.visionengine

import kotlin.math.max
import kotlin.math.min

/** Детекция в нормализованных [0..1] координатах (top-left origin) */
internal data class DetectedRegion(
  val x: Float,
  val y: Float,
  val width: Float,
  val height: Float,
  val score: Float,
  val classIndex: Int,
)

/**
 * Чистый декодер сырых выходов YOLO (без зависимостей от Android/TFLite).
 * Координаты нормализуются адаптивно: значения крупнее 1.5 считаются
 * пикселями входа модели (`inputSize`).
 */
internal object YoloOutputDecoder {
  /** Классический выход: `[C, N]` (каналы первыми) либо `[N, C]`, cx,cy,w,h */
  fun decodeClassic(
    output: Array<FloatArray>,
    inputSize: Int,
    minScore: Float,
  ): List<DetectedRegion> {
    val channelsFirst = output.size < (output.firstOrNull()?.size ?: 0)
    val count = if (channelsFirst) output[0].size else output.size
    val channels = if (channelsFirst) output.size else output[0].size

    fun value(channel: Int, index: Int): Float =
      if (channelsFirst) output[channel][index] else output[index][channel]

    val regions = ArrayList<DetectedRegion>()
    for (i in 0 until count) {
      var score = 0f
      var classIndex = 0
      for (c in 4 until channels) {
        val classScore = value(c, i)
        if (classScore > score) {
          score = classScore
          classIndex = c - 4
        }
      }
      if (score < minScore) {
        continue
      }
      val cx = normalized(value(0, i), inputSize)
      val cy = normalized(value(1, i), inputSize)
      val w = normalized(value(2, i), inputSize)
      val h = normalized(value(3, i), inputSize)
      val x = (cx - w / 2f).coerceIn(0f, 1f)
      val y = (cy - h / 2f).coerceIn(0f, 1f)
      regions.add(
        DetectedRegion(
          x = x,
          y = y,
          width = min(w, 1f - x),
          height = min(h, 1f - y),
          score = score,
          classIndex = classIndex,
        ),
      )
    }

    return regions
  }

  /** End-to-end выход: строки `x1,y1,x2,y2,score,class`, дублей нет */
  fun decodeEndToEnd(
    output: Array<FloatArray>,
    inputSize: Int,
    minScore: Float,
  ): List<DetectedRegion> {
    val regions = ArrayList<DetectedRegion>()
    for (row in output) {
      val score = row[4]
      if (score < minScore) {
        continue
      }
      val x1 = normalized(row[0], inputSize).coerceIn(0f, 1f)
      val y1 = normalized(row[1], inputSize).coerceIn(0f, 1f)
      val x2 = normalized(row[2], inputSize).coerceIn(0f, 1f)
      val y2 = normalized(row[3], inputSize).coerceIn(0f, 1f)
      if (x2 <= x1 || y2 <= y1) {
        continue
      }
      regions.add(
        DetectedRegion(
          x = x1,
          y = y1,
          width = x2 - x1,
          height = y2 - y1,
          score = score,
          classIndex = row[5].toInt(),
        ),
      )
    }
    regions.sortByDescending { it.score }

    return regions
  }

  /**
   * Жадный NMS внутри класса: кандидат с IoU выше порога к уже принятой
   * детекции ТОГО ЖЕ класса отбрасывается. Боксы разных классов друг друга
   * не подавляют — у многоклассовых моделей соседние области (номер, тип,
   * веса) частично перекрываются.
   */
  fun nms(regions: List<DetectedRegion>, iouThreshold: Float): List<DetectedRegion> {
    val sorted = regions.sortedByDescending { it.score }
    val kept = ArrayList<DetectedRegion>()
    for (candidate in sorted) {
      if (kept.none { it.classIndex == candidate.classIndex && iou(it, candidate) > iouThreshold }) {
        kept.add(candidate)
      }
    }

    return kept
  }

  private fun normalized(value: Float, inputSize: Int): Float {
    return if (value > 1.5f) value / inputSize else value
  }

  private fun iou(a: DetectedRegion, b: DetectedRegion): Float {
    val left = max(a.x, b.x)
    val top = max(a.y, b.y)
    val right = min(a.x + a.width, b.x + b.width)
    val bottom = min(a.y + a.height, b.y + b.height)
    if (right <= left || bottom <= top) {
      return 0f
    }
    val intersection = (right - left) * (bottom - top)
    val union = a.width * a.height + b.width * b.height - intersection

    return if (union <= 0f) 0f else intersection / union
  }
}
